pkg/reconciler/gc/gc.go

```diff
@@ -0,0 +1,156 @@
+type Reconciler struct {
+	*reconciler.Base
+
+	// listers index properties about resources
+	configurationLister listers.ConfigurationLister
+	revisionLister      listers.RevisionLister
+
+	configStore reconciler.ConfigStore
+}
+
+// Check that our Reconciler implements controller.Reconciler
+var _ controller.Reconciler = (*Reconciler)(nil)
+
+// Reconcile compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the Garbage Collection
+// resource with the current status of the resource.
+func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
+	// Convert the namespace/name string into a distinct namespace and name.
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		c.Logger.Errorf("invalid resource key: %s", key)
+		return nil
+	}
+	logger := logging.FromContext(ctx)
+
+	ctx = c.configStore.ToContext(ctx)
+
+	// Get the Configuration resource with this namespace/name.
+	original, err := c.configurationLister.Configurations(namespace).Get(name)
+	if errors.IsNotFound(err) {
+		// The resource no longer exists, in which case we stop processing.
+		logger.Errorf("configuration %q in work queue no longer exists", key)
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	// Don't modify the informer's copy.
+	config := original.DeepCopy()
+
+	// Reconcile this copy of the configuration and then write back any status
+	// updates regardless of whether the reconciliation errored out.
+	reconcileErr := c.reconcile(ctx, config)
+	if reconcileErr != nil {
+		c.Recorder.Event(config, corev1.EventTypeWarning, "InternalError", reconcileErr.Error())
+	}
+	return reconcileErr
+}
+
+func (c *Reconciler) reconcile(ctx context.Context, config *v1alpha1.Configuration) error {
+	cfg := configns.FromContext(ctx).RevisionGC
+	logger := logging.FromContext(ctx)
+
+	selector := labels.Set{serving.ConfigurationLabelKey: config.Name}.AsSelector()
+	revs, err := c.revisionLister.Revisions(config.Namespace).List(selector)
+	if err != nil {
+		return err
+	}
+
+	gcSkipOffset := cfg.StaleRevisionMinimumGenerations
+
+	if gcSkipOffset >= int64(len(revs)) {
+		return nil
+	}
+
+	// Sort by creation timestamp descending
+	sort.Slice(revs, func(i, j int) bool {
+		return revs[j].CreationTimestamp.Before(&revs[i].CreationTimestamp)
+	})
+
+	for _, rev := range revs[gcSkipOffset:] {
+		if isRevisionStale(ctx, rev, config) {
+			err := c.ServingClientSet.ServingV1alpha1().Revisions(rev.Namespace).Delete(rev.Name, &metav1.DeleteOptions{})
+			if err != nil {
+				logger.Errorf("Failed to delete stale revision: %v", err)
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func isRevisionStale(ctx context.Context, rev *v1alpha1.Revision, config *v1alpha1.Configuration) bool {
+	cfg := configns.FromContext(ctx).RevisionGC
+	logger := logging.FromContext(ctx)
+
+	if config.Status.LatestReadyRevisionName == rev.Name {
```

Comment:
Well, I guess you can move this to the top, since it does not use the first two variables, so let's save a few cycles.